packages/teraslice/lib/cluster/services/execution.js

```
@@ -398,13 +398,21 @@ module.exports = function module(context, { clusterMasterServer }) {
                         }))
                         .then(() => allocateWorkers(execution, execution.workers)
                             .catch((err) => {
-                            // this is to catch errors of allocateWorkers
-                            // if allocation fails, they are enqueued
-                                logger.error(`Workers failed to be allocated, they will be enqueued, error: ${parseError(err)}`);
+                                if (isNative) {
+                                    // this is to catch errors of allocateWorkers
+                                    // if allocation fails, they are enqueued
+                                    logger.error(`Workers failed to be allocated, they will be enqueued, error: ${parseError(err)}`);
+                                } else {
+                                    // if k8s allocateWorkers fails, the execution fails
+                                    logger.error(`Failured to create k8s worker deployment ${execution.ex_id}, error: ${parseError(err)}`);
+                                    const errMetaData = executionMetaData(null, parseError(err));
+                                    return setExecutionStatus(execution.ex_id, 'failed', errMetaData);
```

If both @jsnoble and @peterdemartini can think about the changes here and let me know if they thing the native case will still work, that would be great.  Specifically, I have some doubts about my call to `return` here ... is that what keeps Workers from being queued as the native comment says?  Regardless, in k8s the job ends up in `failed`, the workers and execution_controllers removed and I get an error message on the `ex` like I want.  I **think** it's right.

---

So there is a little bit of an issue with the code, maybe not. It seems that you are duplicating the error logic of marking it with error etc etc. If you returned a Promise.reject() it would then go to the outer catch which marks the job as failed with the associated error meta data.  Other than that it seems fine

---

Does that imply that what it did in the allocateSlicer catch here:

https://github.com/terascope/teraslice/blob/v0.42.0/packages/teraslice/lib/cluster/services/execution.js#L407

which was before I modified it, should also be rejecting a promise rather than returning?

---

Disregard my previous comment.

Trying `Promise.reject(err)` in the inner `catch` block (for `allocateWorkers`) ... it doesn't seem to be handled correctly.  The job erroneously ends up in the running state, only to exit later when workers fail to connect.  And the error on the `ex` is wrong.  If the worst flaw in this patch is code duplication, then at least it works right.

```
diff --git a/packages/teraslice/lib/cluster/services/execution.js b/packages/teraslice/lib/cluster/services/execution.js
index 41d2129..81aaeac 100644
--- a/packages/teraslice/lib/cluster/services/execution.js
+++ b/packages/teraslice/lib/cluster/services/execution.js
@@ -405,8 +405,7 @@ module.exports = function module(context, { clusterMasterServer }) {
                                 } else {
                                     // if k8s allocateWorkers fails, the execution fails
                                     logger.error(`Failured to create k8s worker deployment ${execution.ex_id}, error: ${parseError(err)}`);
-                                    const errMetaData = executionMetaData(null, parseError(err));
-                                    return setExecutionStatus(execution.ex_id, 'failed', errMetaData);
+                                    Promise.reject(err);
                                 }
                             })))
                     .catch((err) => {
```

---

According to the diff, you added a `Promise.reject(err)` but never returned it, so it won't be propagated up.

I think we should avoid adding more conditional clustering logic to the execution service if possible